### PR TITLE
Forward

### PR DESCRIPTION
**Pull request by ivolarevic**
# Description

Implemented forward functionality with search.

image
image
image

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Dev testing.

**Test Configuration**:

* Firmware version: G930FXXS7ETA1
* Hardware: Samsung galaxy s7
* SDK: Android 8, Oreo

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: app/src/main/java/com/clover/studio/spikamessenger/data/models/networking/responses/ForwardMessagesResponse.kt [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    val status: String,
+    val data: ForwardMessages,
+)
+
+data class ForwardMessages(
+    val messages: List<Message>,
+    val newRooms: List<ChatRoom>
+)
+
+//data class NewRooms(
```
Comment (MatkoTom) [inline app/src/main/java/com/clover/studio/spikamessenger/data/models/networking/responses/ForwardMessagesResponse.kt:16]: If these are not used anymore, feel free to delete them.

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatAdapter.kt [hunk 2]
```diff
@@ -493,6 +500,12 @@ class ChatAdapter(
...
                 }
 
+                holder.binding.tvForward.visibility = if (it.message.isForwarded){
+                    View.VISIBLE
+                } else {
+                    View.GONE
+                }
+
                 /** Show edited layout: */
                 if (it.message.deleted == false && it.message.createdAt != it.message.modifiedAt) {
```
Comment (MatkoTom) [inline app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatAdapter.kt:510]: You could have done the same update, as above ,here

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatMessagesFragment.kt [hunk 3]
```diff
@@ -1024,6 +1038,41 @@ class ChatMessagesFragment : BaseFragment() {
...
+
+                        val users = JsonArray()
+                        userIds.forEach { user ->
+                            users.add(user)
+                        }
+
+                        val message = JsonArray()
+                        message.add(msg.message.id)
+
+                        jsonObject.add("roomIds", rooms)
```
Comment (MatkoTom) [inline app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatMessagesFragment.kt:1062]: Move these fields to Const.kt

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/contacts/ContactsFragment.kt [hunk 4]
```diff
@@ -30,10 +29,10 @@ import timber.log.Timber
```
Comment (MatkoTom) [inline app/src/main/java/com/clover/studio/spikamessenger/ui/main/contacts/ContactsFragment.kt:29]: Do we need to use the same methodology as forwarding here? First submitting "null" to the adapter and then the new list when searching?

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/create_room/NewRoomFragment.kt [hunk 5]
```diff
@@ -38,12 +38,12 @@ import kotlin.streams.toList
 class NewRoomFragment : BaseFragment() {
```
Comment (MatkoTom) [inline app/src/main/java/com/clover/studio/spikamessenger/ui/main/create_room/NewRoomFragment.kt:38]: Same question here as in ContactsFragment. Should we consider how this submitList is working?

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/rooms/RoomsFragment.kt [hunk 6]
```diff
@@ -385,4 +385,9 @@ class RoomsFragment : BaseFragment() {
         binding.topAppBar.menu.findItem(R.id.search_menu_icon).collapseActionView()
         searchQuery = ""
     }
+
+    override fun onResume() {
```
Comment (MatkoTom) [inline app/src/main/java/com/clover/studio/spikamessenger/ui/main/rooms/RoomsFragment.kt:389]: Please remove this :D

File: app/src/main/res/layout/bottom_sheet_forward.xml [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        android:textSize="@dimen/eighteen_sp_text"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent" />
+
+    <ProgressBar
+        android:id="@+id/pb_forward"
+        android:layout_width="wrap_content"
+        android:layout_height="120dp"
+        android:layout_gravity="center"
+        android:layout_marginTop="16dp"
```
Comment (MatkoTom) [inline app/src/main/res/layout/bottom_sheet_forward.xml:29]: This is a constant, so use it.

File: app/src/main/res/layout/item_user_forward.xml [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content">
+
+    <TextView
+        android:id="@+id/tv_name_placeholder"
+        android:layout_width="0dp"
+        android:layout_height="wrap_content"
+        android:layout_marginStart="8dp"
```
Comment (MatkoTom) [inline app/src/main/res/layout/item_user_forward.xml:11]: Use constants here please

File: app/src/main/res/layout/item_user_forward.xml [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        app:layout_constraintBottom_toBottomOf="parent"
+        app:layout_constraintEnd_toEndOf="@+id/tv_name_placeholder"
+        app:layout_constraintStart_toStartOf="@+id/tv_name_placeholder"
+        app:layout_constraintTop_toBottomOf="@+id/tv_name_placeholder" />
+
+    <com.google.android.material.imageview.ShapeableImageView
+        android:id="@+id/iv_room_user_image"
+        android:layout_width="56dp"
+        android:layout_height="56dp"
+        android:layout_marginTop="8dp"
```
Comment (MatkoTom) [inline app/src/main/res/layout/item_user_forward.xml:39]: Same here

File: app/src/main/res/layout/item_user_forward.xml [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+        app:layout_constraintBottom_toBottomOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent"
+        app:shapeAppearanceOverlay="@style/circle" />
+
+    <ImageView
+        android:id="@+id/iv_send_button"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginEnd="16dp"
```
Comment (MatkoTom) [inline app/src/main/res/layout/item_user_forward.xml:53]: Same here

File: app/src/main/java/com/clover/studio/spikamessenger/data/models/entity/ChatRoom.kt [hunk 11]
```diff
@@ -57,4 +57,12 @@ data class ChatRoom @JvmOverloads constructor(
     @Ignore
     @IgnoredOnParcel
     val hasAvatar: Boolean = avatarFileId != null && avatarFileId!! > 0L
+
+    @Ignore
+    @IgnoredOnParcel
```
Comment (MatkoTom) [inline app/src/main/java/com/clover/studio/spikamessenger/data/models/entity/ChatRoom.kt:62]: Remove this if not needed

File: app/src/main/java/com/clover/studio/spikamessenger/data/models/entity/User.kt [hunk 12]
```diff
@@ -63,4 +62,9 @@ data class User(
     val formattedDisplayName: String
         get() = displayName?.takeIf { it.isNotEmpty() }
             ?: MainApplication.appContext.getString(R.string.unknown_user)
+
+    @IgnoredOnParcel
+    @Ignore
```
Comment (MatkoTom) [inline app/src/main/java/com/clover/studio/spikamessenger/data/models/entity/User.kt:67]: Remove this if not needed

**Review by MatkoTom — CHANGES_REQUESTED** [review]
Please review the comments and make the ncessary adjustments

File: app/src/main/java/com/clover/studio/spikamessenger/data/models/networking/responses/ForwardMessagesResponse.kt (hunk 1 above)
Comment (ivolarevic) [reply app/src/main/java/com/clover/studio/spikamessenger/data/models/networking/responses/ForwardMessagesResponse.kt:16]: Done

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatAdapter.kt (hunk 2 above)
Comment (ivolarevic) [reply app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatAdapter.kt:510]: Done

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatMessagesFragment.kt (hunk 3 above)
Comment (ivolarevic) [reply app/src/main/java/com/clover/studio/spikamessenger/ui/main/chat/ChatMessagesFragment.kt:1062]: Done

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/rooms/RoomsFragment.kt (hunk 6 above)
Comment (ivolarevic) [reply app/src/main/java/com/clover/studio/spikamessenger/ui/main/rooms/RoomsFragment.kt:389]: Done xD

File: app/src/main/res/layout/bottom_sheet_forward.xml (hunk 7 above)
Comment (ivolarevic) [reply app/src/main/res/layout/bottom_sheet_forward.xml:29]: Done

File: app/src/main/res/layout/item_user_forward.xml (hunk 8 above)
Comment (ivolarevic) [reply app/src/main/res/layout/item_user_forward.xml:11]: Done

File: app/src/main/java/com/clover/studio/spikamessenger/data/models/entity/ChatRoom.kt (hunk 11 above)
Comment (ivolarevic) [reply app/src/main/java/com/clover/studio/spikamessenger/data/models/entity/ChatRoom.kt:62]: Done

File: app/src/main/java/com/clover/studio/spikamessenger/data/models/entity/User.kt (hunk 12 above)
Comment (ivolarevic) [reply app/src/main/java/com/clover/studio/spikamessenger/data/models/entity/User.kt:67]: Done

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/contacts/ContactsFragment.kt (hunk 4 above)
Comment (ivolarevic) [reply app/src/main/java/com/clover/studio/spikamessenger/ui/main/contacts/ContactsFragment.kt:29]: Done

File: app/src/main/java/com/clover/studio/spikamessenger/ui/main/create_room/NewRoomFragment.kt (hunk 5 above)
Comment (ivolarevic) [reply app/src/main/java/com/clover/studio/spikamessenger/ui/main/create_room/NewRoomFragment.kt:38]: Done

**Review by MatkoTom — APPROVED** [review]
LGTM

**Pull request merged**